cpp/test/neighbors/ann_vamana.cuh

```diff
@@ -0,0 +1,219 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "../test_utils.cuh"
+#include "ann_utils.cuh"
+#include <raft/core/resource/cuda_stream.hpp>
+
+#include "naive_knn.cuh"
+
+#include <cuvs/distance/distance.hpp>
+#include <cuvs/neighbors/vamana.hpp>
+#include <raft/core/device_mdspan.hpp>
+#include <raft/core/device_resources.hpp>
+#include <raft/core/host_mdarray.hpp>
+#include <raft/core/host_mdspan.hpp>
+#include <raft/core/logger.hpp>
+#include <raft/linalg/add.cuh>
+#include <raft/random/rng.cuh>
+#include <raft/util/itertools.hpp>
+
+#include <rmm/device_buffer.hpp>
+
+#include <gtest/gtest.h>
+
+#include <thrust/sequence.h>
+
+#include <cstddef>
+#include <iostream>
+#include <optional>
+#include <string>
+#include <vector>
+
+namespace cuvs::neighbors::vamana {
+
+struct AnnVamanaInputs {
+  int n_rows;
+  int dim;
+  int graph_degree;
+  int visited_size;
+  double max_fraction;
+  cuvs::distance::DistanceType metric;
+  bool host_dataset;
+};
+
+template <typename DataT, typename IdxT>
+inline void CheckGraph(vamana::index<DataT, IdxT>* index_,
+                       AnnVamanaInputs inputs,
+                       cudaStream_t stream)
+{
+  EXPECT_TRUE(index_->graph().size() == (inputs.n_rows * inputs.graph_degree));
+  EXPECT_TRUE(index_->graph().extent(0) == inputs.n_rows);
+  EXPECT_TRUE(index_->graph().extent(1) == inputs.graph_degree);
+
+  // Copy graph to host
+  auto h_graph = raft::make_host_matrix<IdxT, int64_t>(inputs.n_rows, inputs.graph_degree);
+  raft::copy(h_graph.data_handle(), index_->graph().data_handle(), index_->graph().size(), stream);
+
+  size_t edge_count = 0;
+  int max_degree    = 0;
+  for (int i = 0; i < h_graph.extent(0); i++) {
+    int temp_degree = 0;
+    for (int j = 0; j < h_graph.extent(1); j++) {
+      if (h_graph(i, j) < (uint32_t)(inputs.n_rows)) temp_degree++;
+    }
+    if (temp_degree > max_degree) max_degree = temp_degree;
+    edge_count += (size_t)temp_degree;
+  }
+
+  // Tests for acceptable range of edges - low dim can also impact this
+  // Minimum expected maximum degree across the whole graph
+  EXPECT_TRUE(max_degree >= std::min(inputs.graph_degree, inputs.dim));
+
+  float max_edges = (float)(inputs.n_rows * std::min(inputs.graph_degree, inputs.dim));
+
+  RAFT_LOG_INFO("dim:%d, degree:%d, visited_size:%d, edge_count:%lu, max_edges:%lu\n",
+                inputs.dim,
+                inputs.graph_degree,
+                inputs.visited_size,
+                edge_count,
+                (size_t)max_edges);
+
+  // Graph won't always be full, but <75% is very unlikely
+  EXPECT_TRUE(((float)edge_count / max_edges) > 0.75);
+
+  // TODO - Anything else we can test without search
```

Comment:
Until we have better, can't we do a search test using CAGRA? E.g.:
```suggestion
  // Replace invalid edges with an edge to node 0
  auto graph_valid = raft::make_device_vector<unt32_t>(res, index_->graph.size());
  struct edge_op {
  template <typename Type, typename... UnusedArgs>
    constexpr RAFT_INLINE_FUNCTION auto operator()(const Type& in, UnusedArgs...) const 
    { 
      return in == raft::upper_bound<IdxT>() ? Type(0) : in;
    }
   };
  raft::linalg::map(res, graph_valid.view(), edge_op{}, index_->graph());
  
  auto cagra_index = cagra::index<T, IdxT>(res, params.metric, dataset, cagra_graph.view());        
  cagra::search(...);
```

Reply:
There may be some compatability issues between CAGRA search and the vamana graph as it is represented. Need to do a little testing to determine if this is doable currently.

Reply:
I do not expect this to be perfect, so I would definitely not set tight recall tolerance. But for the small datasets used in the example I would expect that this would work.

Reply:
Added CAGRA search to the tests for all graphs with degree <= 128 (cagra cannot search graphs with degree >= 256).